File: packages/typescript/src/types/markets.ts
import type { PaginationParams, SortField } from './common.js';

/**
 * Market status values.
 */
export type MarketStatus =
  | 'initialized'
  | 'active'
  | 'inactive'
  | 'closed'
  | 'determined'
  | 'finalized';

/**
 * Market result - empty string when not yet determined.
 */
export type MarketResult = 'yes' | 'no' | '';

/**
 * Redemption status for a market.
 */
export type RedemptionStatus = 'open' | 'closed';

/**
 * Market account information including mint addresses and redemption status.
 */
export interface MarketAccount {
  /** YES outcome token mint address */
  yesMint: string;
  /** NO outcome token mint address */
  noMint: string;
  /** Market ledger address */
  marketLedger: string;
  /** Redemption status (open or closed) */
  redemptionStatus: RedemptionStatus;
  /** Scalar outcome percentage (for scalar markets) */
  scalarOutcomePct?: number;
}

/**
 * Market data returned from the API.
 * Represents a single trading instrument within an event.
 */
export interface Market {
  /** Unique ticker identifier for the market */
  ticker: string;
  /** Title of the market */
  title: string;
  /** Subtitle of the market */
  subtitle: string;
  /** Event ticker this market belongs to */
  eventTicker: string;
  /** Current market status */
  status: MarketStatus;
  /** Market result (yes, no, or empty if not determined) */
  result: MarketResult;
  /** Market type description */
  marketType: string;
  /** Subtitle for the YES outcome */
  yesSubTitle: string;
  /** Subtitle for the NO outcome */
  noSubTitle: string;
  /** Whether the market can close early */
  canCloseEarly: boolean;
  /** Primary rules for the market */
  rulesPrimary: string;
  /** Total volume traded */
  volume: number;
  /** Open interest */
  openInterest: number;
  /** Open time (Unix timestamp in seconds) */
  openTime: number;
  /** Close time (Unix timestamp in seconds) */
  closeTime: number;
  /** Expiration time (Unix timestamp in seconds) */
  expirationTime: number;
  /** Market accounts by key */
  accounts: Record<string, MarketAccount>;
  /** Secondary rules (optional) */
  rulesSecondary?: string | null;
  /** Condition for early close (optional) */
  earlyCloseCondition?: string | null;
  /** YES ask price as string (optional) */
  yesAsk?: string | null;
  /** YES bid price as string (optional) */
  yesBid?: string | null;
  /** NO ask price as string (optional) */
  noAsk?: string | null;
  /** NO bid price as string (optional) */
  noBid?: string | null;
}

/**
 * Parameters for fetching markets.
 */
export interface MarketsParams extends PaginationParams {
  /** Filter markets by status */
  status?: MarketStatus;
  /** Filter markets that are initialized (have a corresponding market ledger) */
  isInitialized?: boolean;
  /** Sort field for results */
  sort?: SortField;
}

/**
 * Response from the markets endpoint.
 */
export interface MarketsResponse {
  /** Pagination cursor for next page (number of markets to skip) */
  cursor?: number | null;
  /** Array of markets */
  markets: Market[];
}

/**
 * Parameters for batch fetching markets.
 */
export interface MarketsBatchParams {
  /** Array of market tickers to fetch */
  tickers?: string[];
  /** Array of mint addresses to fetch */
  mints?: string[];
}

/**
 * Response from the markets batch endpoint.
 */
export interface MarketsBatchResponse {
  /** Array of markets */
  markets: Market[];
  /** Pagination cursor (optional) */
  cursor?: number | null;
}

/**
 * Parameters for fetching outcome mints.
 */
export interface OutcomeMintsParams {
  /** 
   * Minimum close timestamp (Unix timestamp in seconds).
   * Only markets with close_time >= minCloseTs will be included.
   */
  minCloseTs?: number;
}

/**
 * Response from the outcome mints endpoint.
 */
export interface OutcomeMintsResponse {
  /** Array of all outcome mints from supported markets */
  mints: string[];
}

/**
 * Parameters for filtering outcome mints.
 */
export interface FilterOutcomeMintsParams {
  /** List of token addresses to filter (max 200) */
  addresses: string[];
}

/**
 * Response from the filter outcome mints endpoint.
 */
export interface FilterOutcomeMintsResponse {
  /** List of addresses that are outcome mints */
  outcomeMints: string[];
}

File: packages/python/src/dflow/types/markets.py
"""Market types for DFlow SDK."""

from typing import Literal

from pydantic import BaseModel, Field

MarketStatus = Literal[
    "initialized", "active", "inactive", "closed", "determined", "finalized"
]
MarketResult = Literal["yes", "no", ""]
RedemptionStatus = Literal["open", "closed", "pending"]


class MarketAccount(BaseModel):
    """Market account information including mint addresses and redemption status."""

    yes_mint: str = Field(alias="yesMint")
    no_mint: str = Field(alias="noMint")
    market_ledger: str = Field(alias="marketLedger")
    redemption_status: RedemptionStatus = Field(alias="redemptionStatus")
    scalar_outcome_pct: int | None = Field(default=None, alias="scalarOutcomePct")

    model_config = {"populate_by_name": True}


class Market(BaseModel):
    """Prediction market data.

    Represents a single trading instrument within an event.
    """

    ticker: str
    title: str
    subtitle: str
    event_ticker: str = Field(alias="eventTicker")
    status: MarketStatus
    result: MarketResult
    market_type: str = Field(alias="marketType")
    yes_sub_title: str = Field(alias="yesSubTitle")
    no_sub_title: str = Field(alias="noSubTitle")
    can_close_early: bool = Field(alias="canCloseEarly")
    rules_primary: str = Field(alias="rulesPrimary")
    volume: float
    liquidity: float | None = None
    open_interest: float = Field(alias="openInterest")
    open_time: int = Field(alias="openTime")
    close_time: int = Field(alias="closeTime")
    expiration_time: int = Field(alias="expirationTime")
    accounts: dict[str, MarketAccount]

    # Optional fields
    rules_secondary: str | None = Field(default=None, alias="rulesSecondary")
    early_close_condition: str | None = Field(default=None, alias="earlyCloseCondition")
    yes_ask: str | None = Field(default=None, alias="yesAsk")
    yes_bid: str | None = Field(default=None, alias="yesBid")
    no_ask: str | None = Field(default=None, alias="noAsk")
    no_bid: str | None = Field(default=None, alias="noBid")

    model_config = {"populate_by_name": True}

    # Backwards-compatible properties
    @property
    def yes_price(self) -> float | None:
        """Get YES price (uses yes_bid as the price)."""
        if self.yes_bid:
            return float(self.yes_bid)
        return None

    @property
    def no_price(self) -> float | None:
        """Get NO price (uses no_bid as the price)."""
        if self.no_bid:
            return float(self.no_bid)
        return None

    @property
    def rules(self) -> str:
        """Backwards-compatible alias for rules_primary."""
        return self.rules_primary


class MarketsParams(BaseModel):
    """Parameters for fetching markets."""

    # Filter markets by status
    status: MarketStatus | None = None
    # Filter markets that are initialized (have a corresponding market ledger)
    is_initialized: bool | None = Field(default=None, alias="isInitialized")
    # Sort field for results
    sort: str | None = None
    # Filter by specific market tickers (comma-separated)
    tickers: str | None = None
    # Filter markets by event ticker
    event_ticker: str | None = Field(default=None, alias="eventTicker")
    # Filter markets by series ticker
    series_ticker: str | None = Field(default=None, alias="seriesTicker")
    # Filter markets closing before this timestamp
    max_close_ts: int | None = Field(default=None, alias="maxCloseTs")
    # Filter markets closing after this timestamp
    min_close_ts: int | None = Field(default=None, alias="minCloseTs")
    # Pagination cursor
    cursor: int | None = None
    # Limit
    limit: int | None = None

    model_config = {"populate_by_name": True}


class MarketsResponse(BaseModel):
    """Response from markets list endpoint."""

    cursor: int | None = None
    markets: list[Market]


class MarketsBatchParams(BaseModel):
    """Parameters for batch market queries."""

    tickers: list[str] | None = None
    mints: list[str] | None = None


class MarketsBatchResponse(BaseModel):
    """Response from batch markets endpoint."""

    markets: list[Market]
    cursor: int | None = None


class OutcomeMintsParams(BaseModel):
    """Parameters for fetching outcome mints."""

    # Minimum close timestamp (Unix timestamp in seconds).
    # Only markets with close_time >= minCloseTs will be included.
    min_close_ts: int | None = Field(default=None, alias="minCloseTs")

    model_config = {"populate_by_name": True}


class OutcomeMintsResponse(BaseModel):
    """Response from the outcome mints endpoint."""

    mints: list[str]


class FilterOutcomeMintsParams(BaseModel):
    """Parameters for filtering outcome mints."""

    # List of token addresses to filter (max 200)
    addresses: list[str]


class FilterOutcomeMintsResponse(BaseModel):
    """Response from filter outcome mints endpoint."""

    outcome_mints: list[str] = Field(alias="outcomeMints")

    model_config = {"populate_by_name": True}
